blog/_src/posts/2025-11-21-parallel-threads.scrbl

```diff
@@ -0,0 +1,675 @@
+           (list 8
+                 (list 4021 4154 0)
+                 (list 4021 4154 2)
+                 (list 5928 776 1)
+                 (list 6006 796 2))]
+
+Of course, most programs are not just simple arithmetic. If we
+change our example to repeatedly convert numbers back and forth
+to strings as we compute Fibonacci @codelink["strfib.rkt"] then
+we can see the effects of the more complex conversions. This version
+also triggers frequent allocation, which lets us see
+how thread-local allocation and parallel garbage collection scale.
+
+@benchmark[@racket[(strfib* 32)]
+           (list 1
+                 (list 197 204 3)
+                 (list 198 205 0)
+                 (list 192 192 0)
+                 (list 203 211 0))
+           (list 4
+                 (list 796 826 10)
+                 (list 780 808 2)
+                 (list 861 222 10)
+                 (list 857 221 10))
+           (list 8
+                 (list 1563 1619 11)
+                 (list 1545 1602 4)
+                 (list 2544 419 59)
+                 (list 2551 406 59))]
+
+From this table we still see reasonable scaling up to four cores,
+but the additional work and the use of the garbage collector limit
+scaling beyond that point.
+
+That first string variant of Fibonacci includes a slight cheat,
+however: it goes out of its way to use a @racket[string->number*]
+wrapper that carefully calls @racket[string->number] in a way that
+avoids evaluating expressions that compute the default values of implicit
+arguments. Computing these implicit arguments involves consulting the
+@tech[#:doc '(lib "scribblings/guide/guide.scrbl")]{parameters}
+@racket[read-decimal-as-inexact] and
+@racket[read-single-flonum] and determining the value of parameters consults the
+current continuation. The precise timing for doing so
+depend on how many frames are in a thread's continuation and, even worse, it also
+immediately blocks a future (so the future waits until @racket[touch]ed,
+eliminating any parallel speedup).
+
+@benchmark[@racket[(strfib 32)]
+           (list 1
+                 (list 751 772 4)
+                 (list 562 578 1)
+                 (list 721 721 1)
+                 (list 851 873 0))
+           (list 4
+                 (list 3085 3169 12)
+                 (list 2303 2364 6)
+                 (list 3103 797 33)
+                 (list 4058 4164 2))
+           (list 8
+                 (list 6225 6409 14)
+                 (list 4608 4730 13)
+                 (list 9166 1493 197)
+                 (list 8135 8353 4))]
+
+The coroutine column shows an improvement because a
+coroutine thread has a smaller continuation than the one in
+the sequential column. Importantly, however, note that the
+absolute time for the coroutine column in this table is more
+than double the absolute time in the previous table, meaning
+that we're not learning anything about parallelism with the
+speed up in this table, but instead the cost of consulting
```

Comment:
```suggestion
speed up in this column, but instead the cost of consulting
```

The x1.3 in this column is not interesting for parallelism, but the x4.0 in the next column is interesting for parallelism.

Reply:
Thanks!